massa-models/src/denunciation.rs

```diff
@@ -0,0 +1,642 @@
+                            .is_ok())
+                } else {
+                    Ok(false)
+                }
+            }
+        }
+    }
+
+    fn is_valid(&self) -> Result<bool, DenunciationError> {
+        let (signature_1, signature_2, hash_1, hash_2, public_key) = match self {
+            Denunciation::Endorsement(de) => {
+                let hash_1 = EndorsementDenunciation::compute_hash_for_sig_verif(
+                    &de.public_key,
+                    &de.slot,
+                    &de.index,
+                    &de.hash_1,
+                )?;
+                let hash_2 = EndorsementDenunciation::compute_hash_for_sig_verif(
+                    &de.public_key,
+                    &de.slot,
+                    &de.index,
+                    &de.hash_2,
+                )?;
+
+                (
+                    de.signature_1,
+                    de.signature_2,
+                    hash_1,
+                    hash_2,
+                    de.public_key,
+                )
+            }
+            Denunciation::BlockHeader(de) => {
+                let hash_1 = BlockHeaderDenunciation::compute_hash_for_sig_verif(
+                    &de.public_key,
+                    &de.slot,
+                    &de.hash_1,
+                )?;
+                let hash_2 = BlockHeaderDenunciation::compute_hash_for_sig_verif(
+                    &de.public_key,
+                    &de.slot,
+                    &de.hash_2,
+                )?;
+
+                (
+                    de.signature_1,
+                    de.signature_2,
+                    hash_1,
+                    hash_2,
+                    de.public_key,
+                )
+            }
+        };
+
+        Ok(hash_1 != hash_2
+            && signature_1 != signature_2
+            && public_key.verify_signature(&hash_1, &signature_1).is_ok()
+            && public_key.verify_signature(&hash_2, &signature_2).is_ok())
+    }
+}
+
+/// Create a new Denunciation from 2 SecureShareEndorsement
+impl TryFrom<(&SecureShareEndorsement, &SecureShareEndorsement)> for Denunciation {
+    type Error = DenunciationError;
+
+    fn try_from(
+        (s_e1, s_e2): (&SecureShareEndorsement, &SecureShareEndorsement),
+    ) -> Result<Self, Self::Error> {
+        // Cannot use the same endorsement twice
+        if s_e1 == s_e2 {
+            return Err(DenunciationError::InvalidInput);
+        }
+
+        // In order to create a Denunciation, there should be the same
+        // slot, index & public key
+        if s_e1.content.slot != s_e2.content.slot
+            || s_e1.content.index != s_e2.content.index
+            || s_e1.content_creator_pub_key != s_e2.content_creator_pub_key
+        {
+            return Err(DenunciationError::InvalidInput);
+        }
+
+        // Check sig of s_e1 with s_e1.public_key, s_e1.slot, s_e1.index
+        let s_e1_hash_content = EndorsementDenunciation::compute_content_hash(&s_e1.content)?;
```

Comment:
we shouldn't have to recompute this, because we have the ID of the object. But maybe it's only for tests (?)

Reply:
Add a TODO so this will be done in a follow up